### PR TITLE
[OCPCLOUD-893] Refactor csr approver to use controller runtime

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
Moving out old controller parts of csr approver, and replacing them with the controller-runtime.

### CODE REVIEW

**Comment by michaelgugino** [comment]
This is failing to run:
```
cluster-machine-approver flag redefined: log_dir
panic: cluster-machine-approver flag redefined: log_dir
```

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -39,8 +39,7 @@ func (mc MetricsCollector) Describe(ch chan<- *prometheus.Desc) {
 
 // Collect implements the prometheus.Collector interface.
 func (mc MetricsCollector) collectMetrics(ch chan<- prometheus.Metric) {
-	pending := recentlyPendingCSRs(mc.cacheIndexer)
-	ch <- prometheus.MustNewConstMetric(CurrentPendingCSRCountDesc, prometheus.GaugeValue, float64(pending))
-	ch <- prometheus.MustNewConstMetric(MaxPendingCSRDesc, prometheus.GaugeValue, float64(atomic.LoadUint32(&maxPendingCSRs)))
+	ch <- prometheus.MustNewConstMetric(CurrentPendingCSRCountDesc, prometheus.GaugeValue, float64(atomic.LoadUint32(&controller.PendingCSRs)))
+	ch <- prometheus.MustNewConstMetric(MaxPendingCSRDesc, prometheus.GaugeValue, float64(atomic.LoadUint32(&controller.MaxPendingCSRs)))
```
Comment (JoelSpeed) [inline pkg/metrics/metrics.go:43]: We really should remove this custom metrics collection logic and just use prometheus gauges and observe the values inline

**Review by JoelSpeed — COMMENTED** [review]
/hold This will need to wait until master opens for 4.8 before we can merge. Once master opens let's make sure it merges quick so it has plenty of time to soak.

Found a couple of issues while reading through that will need to be fixed, I might suggest we look at creating a follow up PR for changing the metrics on this, WDYT? 

Equally, would be good to get a follow up to upgrade controller-runtime to v0.7.0, shouldn't be too complicated but we will want to pass the new context through the reconcile methods so we don't randomly use `context.Background()` everywhere (Nit: should really be using `context.TODO()` in controller code to make sure we eventually use a real context)

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+		}
+	}
+
+	return reconcile.Result{}, fmt.Errorf("Failed to find CSR: %v", req)
+}
+
+// reconcileLimits will short circut logic if number of pending CSRs is exceeding limit
+func reconcileLimits(machines *machinev1.MachineList, csrs *certificatesv1.CertificateSigningRequestList) bool {
+	maxPending := getMaxPending(machines.Items)
+	atomic.StoreUint32(&MaxPendingCSRs, uint32(maxPending))
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply pkg/controller/controller.go:140]: SGTM, but I think it is out of scope for this PR. Will be glad to pick it up in a followup.

**Comment by Danil-Grigorev** [comment]
@JoelSpeed PTAL - feedback is adressed.

File: Makefile [hunk 3]
```diff
@@ -14,7 +14,7 @@ test:
 .PHONY: test
 
 unit:
-	go test -v .
+	go test -v -p 1 ./...
```
Comment (JoelSpeed) [inline Makefile:17]: Do we need to specify the `-p` flag here? We don't do it normally

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+			GenericFunc: func(e event.GenericEvent) bool { return pendingCertFilter(e.Object) },
+			DeleteFunc:  func(e event.DeleteEvent) bool { return false },
+		})).
+		Watches(
+			&source.Kind{Type: &corev1.ConfigMap{}},
+			handler.EnqueueRequestsFromMapFunc(m.toCSRs),
+			builder.WithPredicates(predicate.Funcs{
+				CreateFunc:  func(e event.CreateEvent) bool { return caConfigMapFilter(e.Object, nil) },
+				UpdateFunc:  func(e event.UpdateEvent) bool { return caConfigMapFilter(e.ObjectOld, e.ObjectNew) },
+				GenericFunc: func(e event.GenericEvent) bool { return pendingCertFilter(e.Object) },
```
Comment (JoelSpeed) [inline pkg/controller/controller.go:62]: Should this one be `caConfigMapFilter` too? Potential copy/paste error

File: Makefile (hunk 3 above)
Comment (Danil-Grigorev) [reply Makefile:17]: I noticed occasional unit flakes, which were part of previous test code, due to attempts to operate on a single ClusterOperator resource in parallel, and not allowing it to clean results from previous execution.

Comment (JoelSpeed) [reply Makefile:17]: Is the default not to run tests serially anyway? I had thought this change isn't actually doing anything when I first read it

File: pkg/controller/controller.go (hunk 4 above)
Comment (Danil-Grigorev) [reply pkg/controller/controller.go]: Fixed, thanks @JoelSpeed

**Comment by JoelSpeed** [comment]
Please make sure to squash your commits down to a set of logical atomic commits before we merge, quite a few fixup style commits on the end of this at the moment

File: Makefile (hunk 3 above)
Comment (JoelSpeed) [reply Makefile:17]: Still not sure why this makes a difference, have we declared explicitly that some of our tests are parallel?

Comment (Danil-Grigorev) [reply Makefile:17]: You are right, they are. I just checked. We will probably just need something which ensures the resource was actually deleted. Depends on the frequency of the flakes.

Comment (JoelSpeed) [reply Makefile:17]: Let's remove this for now and fix in a separate PR

**Comment by JoelSpeed** [comment]
/hold cancel

Let's check the tests all pass, but LGTM initially

**Pull request merged**